_delphi_utils_python/delphi_utils/flagging/flag_io.py

```diff
@@ -0,0 +1,369 @@
+    params["resid_end_date"] = r2_date.strftime("%m/%d/%Y")
+
+    es_date = e_date -timedelta(eval_dist)
+    params["eval_start_date"] = es_date.strftime("%m/%d/%Y")
+    params["eval_end_date"] = e_date.strftime("%m/%d/%Y")
+    return params
+
+def raw_df_from_api(flag_p):
+    """Make raw dataframe from covidcast api given parameters."""
+    start_d = pd.to_datetime(flag_p['df_start_date'])
+    end_d = pd.to_datetime(flag_p['df_end_date'])
+    sig = flag_p['sig_str']
+    source = flag_p['sig_fold']
+    all_lags = pd.DataFrame()
+    cols = ['ak', 'al', 'ar', 'as', 'az', 'ca', 'co', 'ct', 'dc', 'de', 'fl', 'ga',
+            'gu', 'hi', 'ia', 'id', 'il', 'in', 'ks', 'ky', 'la',
+            'ma', 'md', 'me', 'mi', 'mn', 'mo', 'mp', 'ms', 'mt', 'nc',
+            'nd', 'ne', 'nh', 'nj', 'nm', 'nv', 'ny', 'oh', 'ok',
+            'or', 'pa', 'pr', 'ri', 'sc', 'sd', 'tn', 'tx', 'ut', 'va', 'vi', 'vt',
+            'wa', 'wi', 'wv', 'wy'
+            ]
+    for lag in flag_p['lags']:
+        tot_df = pd.DataFrame()
+        for date_s in pd.date_range(start_d, end_d):
+            if lag == 'var':
+                data = covidcast.signal(source, sig, date_s, date_s,
+                                        "state")
+            else:
+                data = covidcast.signal(source, sig, date_s, date_s,
+                                    "state", as_of=date_s-timedelta(lag))
+            if data is not None:
+                data = data[['geo_value', 'value', 'time_value']]
+                data.columns=['state', 'value', 'index']
+            else:
+                data = pd.DataFrame()
+                data['state'] = cols
+                data['index'] = date_s
+                data['value'] = None
+            tot_df = pd.concat([tot_df, data])
+        if not tot_df.empty:
+            tot_df = tot_df.set_index(["index", 'state'])
+            tot_df = tot_df.unstack(level=0)
+            tot_df.columns = tot_df.columns.droplevel()
+            tot_df = tot_df.T
+            tot_df['lag'] = lag
+        all_lags = pd.concat([tot_df, all_lags], axis=0)
+
+    if flag_p.get('raw_df', None) is not None:
+        all_lags.to_csv(flag_p['raw_df'])
+    return all_lags
+
+def flagging(params, df_list=None, logger=None):
+    """Organization method for different flagging options."""
+    if logger is None:
+        logger = get_structured_logger(
+            __name__, filename=params["common"].get("log_filename"),
+            log_exceptions=params["common"].get("log_exceptions", True))
+    flag_meta = params['flagging_meta']
+    for i, flag_p in enumerate(params['flagging']):
+        df = None
+        if df_list is not None:
+            df = df_list[i]
+        if df is None:
+            try:
+                df = pd.read_csv(flag_p['raw_df'], index_col=0, parse_dates=[0], header=[0])
+            except FileNotFoundError:
+                df = raw_df_from_api(flag_p)
+        if flag_meta['flagger_type'] == 'flagger_io':
```

Comment:
In other modules, we have a utility function that the params object is handed to first so it can check for valid inputs and fill in defaults if needed. Only after being verified is the params object passed to the main function.